File: packages/agent-python/ghost_doc_agent/sanitize.py
"""
Deep sanitization of arbitrary Python objects before sending to the Hub.
Matching keys are replaced with the string "[REDACTED]".
"""
from __future__ import annotations

import re
from typing import Any

DEFAULT_SANITIZE_KEYS: frozenset[str] = frozenset(
    {
        "password",
        "passwd",
        "token",
        "secret",
        "authorization",
        "api_key",
        "apikey",
        "apitoken",
        "api_token",
        "auth",
        "auth_token",
        "access_token",
        "refresh_token",
        "id_token",
        "bearer",
        "jwt",
        "credential",
        "credentials",
        "private_key",
        "privatekey",
        "client_secret",
        "client_id",
        "session",
        "session_id",
        "sessionid",
        "cookie",
        "set_cookie",
        "x_api_key",
        "ssn",
        "social_security",
        "credit_card",
        "card_number",
        "cvv",
        "pin",
        "bank_account",
        "routing_number",
    }
)

# Regex patterns applied to string *values* to detect secrets regardless of key name.
_SECRET_VALUE_PATTERNS: tuple[re.Pattern[str], ...] = (
    # JWT: three base64url segments separated by dots
    re.compile(r"^[A-Za-z0-9_-]{2,}(?:\.[A-Za-z0-9_-]{2,}){2}$"),
    # Bare credit-card: 13-19 digits with optional spaces/dashes
    re.compile(r"^\d{4}[\s\-]?\d{4}[\s\-]?\d{4}[\s\-]?\d{1,7}$"),
)

_REDACTED = "[REDACTED]"
_CIRCULAR = "[Circular]"


def sanitize_deep(
    value: Any,
    keys: frozenset[str] | None = None,
    *,
    _seen: set[int] | None = None,
) -> Any:
    """
    Recursively walk ``value`` and redact any dict key found in ``keys``.

    - Does NOT mutate the original value.
    - Handles circular references by replacing them with ``"[Circular]"``.
    - ``keys`` defaults to DEFAULT_SANITIZE_KEYS.

    Args:
        value:  The value to sanitize.
        keys:   Set of lowercase key names to redact.
        _seen:  Internal set of object IDs used for cycle detection.
                Do not pass this argument from user code.
    """
    if keys is None:
        keys = DEFAULT_SANITIZE_KEYS

    if _seen is None:
        _seen = set()

    # Primitives — check strings for secret patterns, return others as-is
    if value is None or isinstance(value, (bool, int, float, bytes)):
        return value
    if isinstance(value, str):
        stripped = value.strip()
        if any(p.match(stripped) for p in _SECRET_VALUE_PATTERNS):
            return _REDACTED
        return value

    obj_id = id(value)
    if obj_id in _seen:
        return _CIRCULAR
    _seen.add(obj_id)

    try:
        if isinstance(value, dict):
            return {
                k: _REDACTED
                if k.lower() in keys
                else sanitize_deep(v, keys, _seen=_seen)
                for k, v in value.items()
            }

        if isinstance(value, (list, tuple)):
            result = [sanitize_deep(item, keys, _seen=_seen) for item in value]
            return type(value)(result)  # preserve list vs tuple

        # For other objects, attempt to serialize via __dict__ if present
        if hasattr(value, "__dict__"):
            return sanitize_deep(vars(value), keys, _seen=_seen)

        # Fallback: convert to string representation
        return repr(value)
    finally:
        _seen.discard(obj_id)

File: packages/agent-js/src/sanitize.ts
/** Default set of field names whose values are redacted before sending to the Hub. */
export const DEFAULT_SANITIZE_KEYS: readonly string[] = [
  "password",
  "passwd",
  "token",
  "secret",
  "authorization",
  "api_key",
  "apikey",
  "apitoken",
  "api_token",
  "auth",
  "auth_token",
  "access_token",
  "refresh_token",
  "id_token",
  "bearer",
  "jwt",
  "credential",
  "credentials",
  "private_key",
  "privatekey",
  "client_secret",
  "client_id",
  "session",
  "session_id",
  "sessionid",
  "cookie",
  "set_cookie",
  "x_api_key",
  "ssn",
  "social_security",
  "credit_card",
  "card_number",
  "cvv",
  "pin",
  "bank_account",
  "routing_number",
];

/** Regex patterns applied to string *values* (not keys) to detect secrets. */
const SECRET_VALUE_PATTERNS: readonly RegExp[] = [
  // JWT: three base64url segments separated by dots
  /^[A-Za-z0-9_-]{2,}(?:\.[A-Za-z0-9_-]{2,}){2}$/,
  // Bare credit-card: 13–19 consecutive digits (with optional spaces/dashes)
  /^\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{1,7}$/,
];

/**
 * Custom sanitizer function.
 * Receives the key and value; returns the (possibly redacted) value.
 */
export type SanitizerFn = (key: string, value: unknown) => unknown;

/**
 * - `string[]` — blocklist of key names to redact (case-insensitive)
 * - `SanitizerFn` — custom function called for every key in every object
 */
export type SanitizeConfig = string[] | SanitizerFn;

const REDACTED = "[REDACTED]";
const CIRCULAR = "[Circular]";

/**
 * Deep-clone `value` while sanitizing matching keys.
 * Does not mutate the original value.
 * Handles circular references by replacing them with the string `"[Circular]"`.
 */
export function sanitizeDeep(value: unknown, config: SanitizeConfig): unknown {
  const seen = new WeakSet<object>();
  return walk(value, config, seen);
}

function walk(value: unknown, config: SanitizeConfig, seen: WeakSet<object>): unknown {
  if (value === null || typeof value !== "object") {
    return value;
  }

  if (seen.has(value)) {
    return CIRCULAR;
  }
  seen.add(value);

  if (Array.isArray(value)) {
    const result = value.map((item) => walk(item, config, seen));
    seen.delete(value);
    return result;
  }

  const obj = value as Record<string, unknown>;
  const result: Record<string, unknown> = {};

  for (const key of Object.keys(obj)) {
    const rawValue = obj[key];
    const sanitizedValue = shouldRedact(key, rawValue, config)
      ? REDACTED
      : walk(rawValue, config, seen);
    result[key] = sanitizedValue;
  }

  seen.delete(value);
  return result;
}

function shouldRedact(key: string, value: unknown, config: SanitizeConfig): boolean {
  if (typeof config === "function") {
    return config(key, value) === REDACTED;
  }
  const lowerKey = key.toLowerCase();
  if (config.some((k) => k.toLowerCase() === lowerKey)) return true;
  // Also redact string values that look like secrets (JWT, credit card numbers).
  if (typeof value === "string" && SECRET_VALUE_PATTERNS.some((re) => re.test(value.trim()))) {
    return true;
  }
  return false;
}
